crates/testing/tests/tests_1/upgrade_task_with_vote.rs

```diff
@@ -0,0 +1,190 @@
+    script::{Expectations, TaskScript},
+    view_generator::TestViewGenerator,
+};
+use hotshot_types::{
+    data::{null_block, ViewNumber},
+    simple_vote::UpgradeProposalData,
+    traits::{election::Membership, node_implementation::ConsensusTime},
+    vote::HasViewNumber,
+};
+use vbs::version::Version;
+
+#[cfg(feature = "dependency-tasks")]
+#[cfg_attr(async_executor_impl = "tokio", tokio::test(flavor = "multi_thread"))]
+#[cfg_attr(async_executor_impl = "async-std", async_std::test)]
+/// Tests that we correctly update our internal quorum vote state when reaching a decided upgrade
+/// certificate.
+async fn test_upgrade_task_with_vote() {
+    use hotshot_testing::helpers::build_system_handle;
+
+    async_compatibility_layer::logging::setup_logging();
+    async_compatibility_layer::logging::setup_backtrace();
+
+    let handle = build_system_handle(2).await.0;
+    let quorum_membership = handle.hotshot.memberships.quorum_membership.clone();
+    let da_membership = handle.hotshot.memberships.da_membership.clone();
+
+    let old_version = Version { major: 0, minor: 1 };
+    let new_version = Version { major: 0, minor: 2 };
+
+    let upgrade_data: UpgradeProposalData<TestTypes> = UpgradeProposalData {
+        old_version,
+        new_version,
+        decide_by: ViewNumber::new(6),
+        new_version_hash: [0u8; 12].to_vec(),
+        old_version_last_view: ViewNumber::new(6),
+        new_version_first_view: ViewNumber::new(7),
+    };
+
+    let mut proposals = Vec::new();
+    let mut votes = Vec::new();
+    let mut dacs = Vec::new();
+    let mut vids = Vec::new();
+    let mut leaders = Vec::new();
+    let mut leaves = Vec::new();
+    let consensus = handle.hotshot.consensus().clone();
+    let mut consensus_writer = consensus.write().await;
+
+    let mut generator = TestViewGenerator::generate(quorum_membership.clone(), da_membership);
+    for view in (&mut generator).take(2).collect::<Vec<_>>().await {
+        proposals.push(view.quorum_proposal.clone());
+        votes.push(view.create_quorum_vote(&handle));
+        dacs.push(view.da_certificate.clone());
+        vids.push(view.vid_proposal.clone());
+        leaders.push(view.leader_public_key);
+        leaves.push(view.leaf.clone());
+        consensus_writer.update_validated_state_map(
+            view.quorum_proposal.data.view_number(),
+            build_fake_view_with_leaf(view.leaf.clone()),
+        ).unwrap();
+        consensus_writer.update_saved_leaves(view.leaf.clone());
+    }
+    drop(consensus_writer);
+
+    generator.add_upgrade(upgrade_data);
+
+    for view in generator.take(4).collect::<Vec<_>>().await {
+        proposals.push(view.quorum_proposal.clone());
+        votes.push(view.create_quorum_vote(&handle));
+        dacs.push(view.da_certificate.clone());
+        vids.push(view.vid_proposal.clone());
+        leaders.push(view.leader_public_key);
+        leaves.push(view.leaf.clone());
+    }
+
+    let inputs = vec![
```

Comment:
Same here as above, it should be relatively straightforward to find/replace, then we can guarantee that this works independent of the order.